<a href="https://colab.research.google.com/github/shreyanshchordia/covid19-global-analysis/blob/master/COVID_19.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Viewer is requested to view the [notebook](https://colab.research.google.com/gist/shreyanshchordia/13ea2b892ae1b5ecd6e21ab16d536762/covid-19.ipynb) on google colaboratory. This notebook outputs a bunch of plots and charts that cannot be displayed by github notebook viewer in static mode. Thank you.

#**Loading Libraries**

In [0]:
!pip install pycountry
import pandas as pd
import plotly.express as px
from datetime import datetime
import pycountry
import plotly.graph_objects as go
import numpy as np

#**Loading Datasets**

In [0]:
confirmed = pd.read_csv("/content/drive/My Drive/Colab Notebooks/Corona/confirmed.csv")
death = pd.read_csv("/content/drive/My Drive/Colab Notebooks/Corona/death.csv")
recovered = pd.read_csv("/content/drive/My Drive/Colab Notebooks/Corona/recovered.csv")

In [0]:
confirmed.head()

Province/State Country/Region      Lat  ...  4/5/20  4/6/20  4/7/20
0            NaN    Afghanistan  33.0000  ...     349     367     423
1            NaN        Albania  41.1533  ...     361     377     383
2            NaN        Algeria  28.0339  ...    1320    1423    1468
3            NaN        Andorra  42.5063  ...     501     525     545
4            NaN         Angola -11.2027  ...      14      16      17

[5 rows x 81 columns]

In [0]:
death.head()

Province/State Country/Region      Lat  ...  4/5/20  4/6/20  4/7/20
0            NaN    Afghanistan  33.0000  ...       7      11      14
1            NaN        Albania  41.1533  ...      20      21      22
2            NaN        Algeria  28.0339  ...     152     173     193
3            NaN        Andorra  42.5063  ...      18      21      22
4            NaN         Angola -11.2027  ...       2       2       2

[5 rows x 81 columns]

In [0]:
recovered.head()

Province/State Country/Region      Lat  ...  4/5/20  4/6/20  4/7/20
0            NaN    Afghanistan  33.0000  ...      15      18      18
1            NaN        Albania  41.1533  ...     104     116     131
2            NaN        Algeria  28.0339  ...      90      90     113
3            NaN        Andorra  42.5063  ...      26      31      39
4            NaN         Angola -11.2027  ...       2       2       2

[5 rows x 81 columns]

#**Analysis of Confirmation of Cases over time**

##**Building relevant dataframes**

In [0]:
top_10_confirmed = confirmed.sort_values(by=["4/7/20"],ascending=False,na_position='last')[0:10]
confirmed_flow = top_10_confirmed.melt(id_vars=["Province/State", "Country/Region", "Lat",'Long'],var_name="Date",value_name="Confirmed")
confirmed_flow.head()

Province/State Country/Region      Lat     Long     Date  Confirmed
0            NaN             US  37.0902 -95.7129  1/22/20          1
1            NaN          Spain  40.0000  -4.0000  1/22/20          0
2            NaN          Italy  43.0000  12.0000  1/22/20          0
3            NaN         France  46.2276   2.2137  1/22/20          0
4            NaN        Germany  51.0000   9.0000  1/22/20          0

In [0]:
confirmed_melted = confirmed.melt(id_vars=["Province/State", "Country/Region", "Lat",'Long'],var_name="Date",value_name="Confirmed")
for i,string in enumerate(confirmed_melted['Date']):
    x = datetime.strptime(string,'%m/%d/%y')
    confirmed_melted['Date'][i] = x.strftime('%Y-%m-%d')

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [0]:
confirmed_melted.head()

Province/State Country/Region      Lat     Long        Date  Confirmed
0            NaN    Afghanistan  33.0000  65.0000  2020-01-22          0
1            NaN        Albania  41.1533  20.1683  2020-01-22          0
2            NaN        Algeria  28.0339   1.6596  2020-01-22          0
3            NaN        Andorra  42.5063   1.5218  2020-01-22          0
4            NaN         Angola -11.2027  17.8739  2020-01-22          0

In [0]:
cases = confirmed_melted
grp = cases.groupby(['Date', 'Country/Region'])['Confirmed'].max()
grp = grp.reset_index()
grp['Country'] =  grp['Country/Region']
grp.head()

Date Country/Region  Confirmed      Country
0  2020-01-22    Afghanistan          0  Afghanistan
1  2020-01-22        Albania          0      Albania
2  2020-01-22        Algeria          0      Algeria
3  2020-01-22        Andorra          0      Andorra
4  2020-01-22         Angola          0       Angola

In [0]:
group = confirmed.groupby(['Country/Region'])['4/7/20'].max()
group = group.reset_index()
group = group.sort_values(by=['4/7/20'])
group = group[group['4/7/20']>10000]
group.head()

Country/Region  4/7/20
90    Korea, South   10331
135       Portugal   12442
9          Austria   12639
23          Brazil   14034
120    Netherlands   19580

##**Plots**

In [0]:
fig = px.choropleth(grp, locations="Country", locationmode='country names', 
                     color="Confirmed", hover_name="Country/Region",hover_data = [grp.Confirmed],projection="mercator",
                     animation_frame="Date",width=1000, height=800,
                     color_continuous_scale='blues',
                     range_color=[100,20000],
                     title='World Map of Coronavirus')
fig.update_traces(reversescale=True)
fig.update(layout_coloraxis_showscale=True)
fig.show(renderer="colab")

In [0]:
fig = px.bar(group,y='4/7/20',x='Country/Region',color='4/7/20',title='Spread of corona virus in countries with more than 10,000 cases',
             color_continuous_scale='viridis',range_color=[10000,200000])
fig.update(layout_coloraxis_showscale=True)
fig.show(renderer='colab')

In [0]:
fig = px.pie(confirmed, values='4/7/20', names='Country/Region', color= '4/7/20',title="USA with most number of confirmed cases")
fig.update_traces(textposition='inside', textinfo='percent')
fig.show(renderer="colab")

In [0]:
fig = px.bar(top_10_confirmed,x="Country/Region",y="4/7/20",color="4/7/20",title="Top 10 countries with most number of confirmed cases",
color_continuous_scale='reds',range_color=[1000,150000])
fig.show(renderer="colab")

In [0]:
fig = px.line(confirmed_flow, x="Date", y="Confirmed",color='Country/Region',title="Timeline of confirmation of cases")
fig.update_traces(mode='lines+markers')
fig.show(renderer="colab")

In [0]:
fig = px.bar(confirmed_flow, x="Country/Region", y="Confirmed", color="Confirmed",animation_frame="Date", animation_group="Country/Region", range_y=[0,450000],title="The spread of CORONA?!",
             color_continuous_scale='reds',range_color=[1000,150000])
fig.show(renderer="colab")

#**Analysis of Deaths due to Corona Virus**

##**Building relevant dataframes**

In [0]:
top_10_death = death.sort_values(by=["4/7/20"],ascending=False,na_position='last')[0:10]
death_flow = top_10_death.melt(id_vars=["Province/State", "Country/Region", "Lat",'Long'],var_name="Date",value_name="Dead")


global_deaths = pd.read_csv("/content/drive/My Drive/Colab Notebooks/Corona/total-daily-covid-deaths.csv")

for i,string in enumerate(global_deaths['Date']):
    x = datetime.strptime(string,'%b %d, %Y')
    global_deaths['Date'][i] = x.strftime('%Y-%m-%d')

global_deaths_world = global_deaths[global_deaths['Entity']=='World']
global_deaths_excluding_china = global_deaths[global_deaths['Entity']=='World excl. China']
deaths_china = global_deaths[global_deaths['Entity']=='China']

In [0]:
country_deaths = global_deaths.copy()
result = []
for code in country_deaths['Code']:
    a = pycountry.countries.get(alpha_3=code)
    if a is not None: a = code
    result.append(a)
country_deaths['Code'] = result
country_deaths = country_deaths.dropna()
con_10_total = country_deaths[country_deaths['Total confirmed deaths (deaths)']>=10]

In [0]:
global_deaths.head()

Entity  ... Daily new confirmed deaths (deaths)
0  Afghanistan  ...                                   0
1  Afghanistan  ...                                   0
2  Afghanistan  ...                                   0
3  Afghanistan  ...                                   0
4  Afghanistan  ...                                   0

[5 rows x 5 columns]

In [0]:
global_deaths_world.head()

Entity  ... Daily new confirmed deaths (deaths)
10673  World  ...                                   0
10674  World  ...                                   0
10675  World  ...                                   0
10676  World  ...                                   0
10677  World  ...                                   0

[5 rows x 5 columns]

In [0]:
global_deaths_excluding_china.head()

Entity  ... Daily new confirmed deaths (deaths)
10773  World excl. China  ...                                   0
10774  World excl. China  ...                                   0
10775  World excl. China  ...                                   0
10776  World excl. China  ...                                   0
10777  World excl. China  ...                                   0

[5 rows x 5 columns]

In [0]:
deaths_china.head()

Entity  ... Daily new confirmed deaths (deaths)
2081  China  ...                                   0
2082  China  ...                                   0
2083  China  ...                                   0
2084  China  ...                                   0
2085  China  ...                                   0

[5 rows x 5 columns]

In [0]:
con_10_total.head()

Entity  ... Daily new confirmed deaths (deaths)
89   Afghanistan  ...                                   4
210      Albania  ...                                   1
211      Albania  ...                                   0
212      Albania  ...                                   2
213      Albania  ...                                   3

[5 rows x 5 columns]

##**Plots**

In [0]:
fig = px.bar(top_10_death,x="Country/Region",y="4/7/20",color="4/7/20",title="Top 10 countries with most number of Death cases due to Corona Virus",
             color_continuous_scale='viridis',range_color=[1,20000])
fig.show(renderer="colab")

In [0]:
fig = px.line(death_flow, x="Date", y="Dead", color='Country/Region',title="History of Deaths")
fig.update_traces(mode='lines+markers')
fig.show(renderer="colab")

In [0]:
fig = px.bar(death_flow, x="Country/Region", y="Dead", color="Dead",animation_frame="Date", range_y=[0,20000],title="Deaths with time",
             color_continuous_scale='viridis',range_color=[1,20000])
fig.show(renderer="colab")

In [0]:
fig = go.Figure()
date = global_deaths_world['Date']
total_count_of_deaths = global_deaths_world['Total confirmed deaths (deaths)']
daily_count_of_deaths = global_deaths_world['Daily new confirmed deaths (deaths)']
# Create and style traces
fig.add_trace(go.Scatter(x=date, y=total_count_of_deaths, name='Total confirmed deaths',line=dict(color='firebrick')))
fig.add_trace(go.Scatter(x=date, y=daily_count_of_deaths, name = 'Daily new confirmed deaths',
                         line=dict(color='royalblue')))
# Edit the layout
fig.update_traces(mode='lines+markers')
fig.update_layout(title='Total Deaths worldwide due to corona analysis',
                   xaxis_title='Date',
                   yaxis_title='Deaths')


fig.show(renderer='colab')

In [0]:
fig = go.Figure()
date = global_deaths_world['Date']
y = global_deaths_excluding_china['Total confirmed deaths (deaths)']
z = deaths_china['Total confirmed deaths (deaths)']
# Create and style traces
fig.add_trace(go.Scatter(x=date, y=y, name='World excluding China',line=dict(color='firebrick')))
fig.add_trace(go.Scatter(x=date, y=z, name = 'China',
                         line=dict(color='royalblue')))
# Edit the layout
fig.update_traces(mode='lines+markers')
fig.update_layout(title='World V/S China- Total confirmed Deaths',
                   xaxis_title='Date',
                   yaxis_title='Deaths')


fig.show(renderer='colab')

In [0]:
fig = go.Figure()
date = global_deaths_world['Date']
y = global_deaths_excluding_china['Daily new confirmed deaths (deaths)']
z = deaths_china['Daily new confirmed deaths (deaths)']
# Create and style traces
fig.add_trace(go.Scatter(x=date, y=y, name='World excluding China',line=dict(color='firebrick')))
fig.add_trace(go.Scatter(x=date, y=z, name = 'China',
                         line=dict(color='royalblue')))
# Edit the layout
fig.update_traces(mode='lines+markers')
fig.update_layout(title='World V/S China - Daily deaths',
                   xaxis_title='Date',
                   yaxis_title='Deaths')


fig.show(renderer='colab')

In [0]:
fig = px.line(con_10_total,y="Total confirmed deaths (deaths)", color="Entity",
              width=1000,height=600,range_x=[5,40],range_y=[0,5000])
fig.update_traces(mode='lines+markers',)
fig.update_layout(title='Day wise analysis since the 10th confirmed death is respective countries',
                   xaxis_title='Day wise analysis of countries since their 10th confirmed death',
                   yaxis_title='Deaths')
fig.show(renderer='colab')

#**Analysis of recoveries over time**

In [0]:
top_10_recovered = recovered.sort_values(by=["4/7/20"],ascending=False,na_position='last')[0:10]

recovered_flow = top_10_recovered.melt(id_vars=["Province/State", "Country/Region", "Lat",'Long'],var_name="Date",value_name="Recovered")

##**Plots**

In [0]:
fig = px.bar(top_10_recovered,x="Country/Region",y="4/7/20",color="4/7/20",title="Top 10 Countries with the most number of recoveries")
fig.show(renderer="colab")

In [0]:
fig = px.line(recovered_flow, x="Date", y="Recovered", color='Country/Region',title="The road of recovery")
fig.update_traces(mode='lines+markers')
fig.show(renderer="colab")

In [0]:
fig = px.bar(recovered_flow, x="Country/Region", y="Recovered", color="Recovered",animation_frame="Date", range_y=[0,65000],title="Recovery through time")
fig.show(renderer="colab")

#**Confirmed - Recovered - Dead ?**

##**Building relevant frames**

In [0]:
c = (np.sum(confirmed.values[:,4:],axis=0))
r = (np.sum(recovered.values[:,4:],axis=0))
d = (np.sum(death.values[:,4:],axis=0))
days = [i for i in range(1,78)]
state=[]
for i in days:
    state.append('confirmed cases')
    state.append('recovered cases')
    state.append('death cases')
values=[]
for i in days:
    values.append(c[i-1])
    values.append(r[i-1])
    values.append(d[i-1])
day = []
for i in days:
    day.append(i)
    day.append(i)
    day.append(i)

global_pandemic = pd.DataFrame({'Day':day,'State':state,'Count':values})

In [0]:
fig = px.bar(global_pandemic, x="Count", y="State", color="Count",animation_frame="Day", range_x=[0,1500000],title="Again a worldwide analysis",orientation='h',
             color_continuous_scale='solar',range_color=[10000,600000])
fig.show(renderer="colab")

#**Median Age of countries V/S Death-to-Confirmed Ratio of cases** 

##**Building relevant data frames**

In [0]:
median_age = pd.read_csv("/content/drive/My Drive/Colab Notebooks/Corona/data.csv")
median_age.head()

Place  Median  Median Male  Median Female
0                     Monaco    53.1         51.7           54.5
1                      Japan    47.3         46.0           48.7
2                    Germany    47.1         46.0           48.2
3  Saint Pierre and Miquelon    46.5         46.0           47.0
4                      Italy    45.5         44.4           46.5

In [0]:
df = global_deaths.copy(deep=True)
df = df[df['Date']=='2020-04-08']
result=[]
for name in df['Entity']:
    a = pycountry.countries.get(name=name)
    if a is not None: a = a.alpha_3
    result.append(a)

df['Code'] = result
df = df.dropna()

In [0]:

result=[]
for name in median_age['Place']:
    a = pycountry.countries.get(name=name)
    if a is not None: a = a.alpha_3
    result.append(a)

median_age['CODE'] = result

In [0]:
result = []
for code in df['Code']:
    result.append(np.sum(np.asarray(median_age['Median'])*np.asarray(median_age['CODE']==code)))
df['median']=result

In [0]:
df = df.sort_values(by=['Total confirmed deaths (deaths)'])
df = df[-30:]

In [0]:
top_30_confirmed = confirmed.sort_values(by=["4/7/20"],ascending=True,na_position='last')[-30:]

In [0]:
x = df[['Entity','Total confirmed deaths (deaths)','median']].copy()
y = top_30_confirmed[['Country/Region','4/7/20']].copy()
array = y.values
array[-1][0]='United States'
y['Country/Region'] = list(array[:,0])
new_df = x.merge(y,how='inner',left_on='Entity',right_on ="Country/Region")

In [0]:
array = new_df.values
dead_by_confirm = np.divide(array[:,1],array[:,4])
new_df['dead/confirm ratio'] = list(dead_by_confirm)
new_df = new_df.sort_values(by=['dead/confirm ratio'])
new_df = new_df[new_df['4/7/20']!=9340]
new_df.head()

Entity  Total confirmed deaths (deaths)  ...  4/7/20 dead/confirm ratio
13   Germany                             1861  ...  107663           0.017285
5    Austria                              243  ...   12639           0.019226
12    Turkey                              725  ...   34109           0.021255
0     Poland                              129  ...    4848           0.026609
6   Portugal                              345  ...   12442           0.027729

[5 rows x 6 columns]

##**Plots**

In [0]:
fig = px.bar(new_df, x='Entity', y='dead/confirm ratio',
             hover_data=['dead/confirm ratio', 'Total confirmed deaths (deaths)','4/7/20','median'], color='median', height=400,title='dead/confrim ratio V/S median bar chart')
fig.show(renderer='colab')

In [0]:
fig = px.scatter(new_df, x="median", y="dead/confirm ratio", color="dead/confirm ratio", hover_data=['Entity','median','dead/confirm ratio'],
                 color_continuous_scale='viridis')
fig.update_traces(mode='markers', marker_size=40)
fig.update_layout(title='Better visualisation of "dead/confirm ratio V/S median"',
                  yaxis_zeroline=False, xaxis_zeroline=False)

fig.show(renderer='colab')

#**Conclusion**

The Corona Virus is going to be the cause of recession in the world. Almost entire world is at hault today. Within a span of approximately 5 months, globally 1.5 Million people have got the virus, and approximately 100 thousand people have died. Countries like Italy, Spain and USA seem to have suffered the most. No signs of hope are visible. People are losing their families, friends and loved ones. Countries are falling short of graveyards to bury. 

The analysis also leads to many questions as to how China didn't get affected by the virus to a huge extent despite being the native ground to the virus. How come entire world is under suffering but China has got over it? Germany is among the countries with the most number of confirmed cases yet the mortality rate is low there. What brings the country to this safer side, despite being surrounded by the other European Countries struggling hard to get through. Data analysis and research say that Corona Virus can only be controlled by increasing the amount of testings. Because a person witnesses symptoms of the virus approximately 15-20 days after getting exposed to the virus. Hence for those 15-20 days, the person becomes a carrier of the virus. If we are able to do testings at a faster rate, we can identify carriers, well before they start showing symptoms and hence prevent the widespread. 

This is our duty as data analyists and Machine Learning practioners to find patterns among the data and try to make predictions and solutions. 

#**WE ALL ARE TOGETHER IN THIS FIGHT. STAY HOME ! STAY SAFE !**